services/listen_dnsport.c

```diff
@@ -1153,30 +1191,412 @@ make_sock_port(int stype, const char* ifname, int port,
+
+	data->token = token;
+
+	coap_opt_iterator_t opt_iter;
+	coap_option_iterator_init((coap_pdu_t*)pdu, &opt_iter, COAP_OPT_ALL);
+	coap_opt_t* option;
+	while ((option = coap_option_next(&opt_iter))) {
+		coap_insert_optlist(&data->options, coap_new_optlist(opt_iter.number, coap_opt_length(option), coap_opt_value(option)));
+	}
+}
+
+static void
+doc_handle_fetch(coap_resource_t *resource, coap_session_t *session,
+		const coap_pdu_t *request, const coap_string_t *query,
+		coap_pdu_t *response) {
+
+	struct comm_point* cp = coap_resource_get_userdata(resource);
+
+	size_t size;
+	const uint8_t* data;
+	struct comm_reply rep;
+	struct coap_pdu_response_data* pdu_wrapper = (struct coap_pdu_response_data*)malloc(sizeof(struct coap_pdu_response_data));
+	rep.c = (struct comm_point*)cp;
+	rep.session = session;
+	rep.response = response;
+
+	struct sldns_buffer* buffer;
+
+	doc_extract_pdu_info(response, pdu_wrapper);
+
+	rep.pdu_wrapper = pdu_wrapper;
+
+	/* Get remote address for the CoAP session */
+	const coap_address_t* remote_addr = coap_session_get_addr_remote(session);
+
+	/* Ensure that remote IPv4 or IPv6 addresses are set */
+	if (remote_addr->addr.sa.sa_family == AF_INET) {
+		/* Copy IPv4 address */
+		memcpy(&rep.remote_addr, &remote_addr->addr.sin, sizeof(struct sockaddr_in));
+		rep.remote_addrlen = sizeof(struct sockaddr_in);
+		memcpy(&rep.client_addr, &remote_addr->addr.sin, sizeof(struct sockaddr_in));
+		rep.client_addrlen= sizeof(struct sockaddr_in);
+	} else if (remote_addr->addr.sa.sa_family == AF_INET6) {
+		/* Copy IPv6 address */
+		memcpy(&rep.remote_addr, &remote_addr->addr.sin6, sizeof(struct sockaddr_in6));
+		rep.remote_addrlen = sizeof(struct sockaddr_in6);
+		memcpy(&rep.client_addr, &remote_addr->addr.sin, sizeof(struct sockaddr_in6));
+		rep.client_addrlen= sizeof(struct sockaddr_in6);
+	}
+
+	if (!coap_get_data(request, &size, &data)) {
+		/* TODO error handling? */
+	}
+
+	/* Ensure that the buffer of the comm_point is large enough */
+	if (sldns_buffer_capacity(rep.c->buffer) < size) {
+		/* TODO expand buffer if necessary */
+	}
+
+	sldns_buffer_clear(rep.c->buffer);
+
+	rep.remote_addrlen = (socklen_t)sizeof(rep.remote_addr);
+
+	sldns_buffer_write(rep.c->buffer, data, size);
+	sldns_buffer_flip(rep.c->buffer);
+
+	rep.srctype = 0;
+	rep.is_proxied = 0;
+
+	fptr_ok(fptr_whitelist_comm_point(rep.c->callback));
+	if ((*rep.c->callback)(rep.c, rep.c->cb_arg, NETEVENT_NOERROR, &rep)) {
+		buffer = rep.c->buffer;
+		const uint8_t* buffer_data = sldns_buffer_begin(buffer);
+		size_t buffer_length =sldns_buffer_remaining(buffer);
+		coap_pdu_set_code(response, COAP_RESPONSE_CODE_CONTENT);
+
+		coap_add_data(response, buffer_length, (const uint8_t*)buffer_data);
+	}
+}
+
+static void
+doc_init_resources(coap_context_t* ctx, const char* resource_path, struct comm_point* cp) {
+	coap_resource_t* r;
+
+	log_info("Registering coap resource `%s`\n", resource_path);
```

Comment:
Maybe this information is only needed when verbosity is higher.
```suggestion
	verbose(VERB_DETAIL, "Registering coap resource `%s`\n", resource_path);
```